File: scripts/oop-practice-sources/php/shop-cart.php
<?php

class Product
{
    public function __construct(
        public string $name,
        public int $price
    ) {}
}

class Cart
{
    /** @var list<Product> */
    public array $items = [];

    public function add(Product $product): void
    {
        $this->items[] = $product;
        echo "В корзину добавлено: {$product->name} ({$product->price} ₽)\n";
    }

    public function total(): int
    {
        return array_sum(array_map(fn(Product $p) => $p->price, $this->items));
    }
}

class Order
{
    /** @var list<Product> */
    public array $items;
    public int $total;

    public function __construct(Cart $cart)
    {
        $this->items = $cart->items;
        $this->total = $cart->total();
    }

    public function checkout(): void
    {
        echo "Оформление заказа...\n";
        foreach ($this->items as $item) {
            echo "  — {$item->name}: {$item->price} ₽\n";
        }
        echo "Итого: {$this->total} ₽\n";
        echo "Заказ оформлен!\n";
    }
}

$cart = new Cart();
$cart->add(new Product('Книга', 500));
$cart->add(new Product('Ручка', 50));
$order = new Order($cart);
$order->checkout();

File: scripts/oop-practice-sources/csharp/shop-cart.cs
class Product
{
    public string Name { get; }
    public int Price { get; }

    public Product(string name, int price)
    {
        Name = name;
        Price = price;
    }
}

class Cart
{
    public List<Product> Items { get; } = new();

    public void Add(Product product)
    {
        Items.Add(product);
        Console.WriteLine($"В корзину добавлено: {product.Name} ({product.Price} ₽)");
    }

    public int Total()
    {
        return Items.Sum(p => p.Price);
    }
}

class Order
{
    public List<Product> Items { get; }
    public int Total { get; }

    public Order(Cart cart)
    {
        Items = new List<Product>(cart.Items);
        Total = cart.Total();
    }

    public void Checkout()
    {
        Console.WriteLine("Оформление заказа...");
        foreach (var item in Items)
        {
            Console.WriteLine($"  — {item.Name}: {item.Price} ₽");
        }
        Console.WriteLine($"Итого: {Total} ₽");
        Console.WriteLine("Заказ оформлен!");
    }
}

class Program
{
    static void Main()
    {
        var cart = new Cart();
        cart.Add(new Product("Книга", 500));
        cart.Add(new Product("Ручка", 50));
        var order = new Order(cart);
        order.Checkout();
    }
}
